### PR TITLE
pkg/kv/kvclient: update per-vCPU concurrency limits

### PR DESCRIPTION
**Pull request by sean-**
Increase the current values (arbitrarily selected) by 6x (also an
arbitrary increase).  Annecdotal evidence from customers suggests
this may unlock some workloads where vCPU utilization appears to
"plateau" around 70%, and now allows for ~100% vCPU utilization.
Until https://github.com/cockroachdb/cockroach/issues/131125 and https://github.com/cockroachdb/cockroach/issues/131126 are implemented, it is not possible to
directly monitor these incremental improvements to the defaults.

Epic: CRDB-42388
Epic: CRDB-42389

Release note (performance improvement): Increase the per-vCPU
concurrency limits for KV operations.  Specifically, increase
`kv.dist_sender.concurrency_limit` to 384/vCPU (up from 64/vCPU)
and `kv.streamer.concurrency_limit` to 96/vCPU (up from 8/vCPU).

### CODE REVIEW

**Review by nvb — COMMENTED** [review]
Have you run any additional tests with this to demonstrate that it is (1) beneficial, (2) not harmful to mixed workloads?

Also, the change only has an effect for nodes with fewer than 16 vCPUs, because the maximum values have not changed. Is that intentionally?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131226#-:-O7VVCAT5buH9Rc2ZfFd:b-wh810o)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @sean-)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 348 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131226#-O7VT2AQC-gq44YrS1AC:-O7VT2AQC-gq44YrS1AD:b-d2raod) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d8998de678cb3d7bde2cae6fb5303e44c5f05ff/pkg/kv/kvclient/kvcoord/dist_sender.go#L348)):*
> ```Go
> 	// The default limit for asynchronous senders.
> 	defaultSenderConcurrency = 1024
> 	// The default scaling factor for the number of async ops per vCPU
> ```

Missing punctuation.
___
*[`pkg/kv/kvclient/kvstreamer/streamer.go` line 346 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131226#-O7VT5WZ1X10yUF7MJic:-O7VT5W_ByOAK3tRdQtD:b-d2raod) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d8998de678cb3d7bde2cae6fb5303e44c5f05ff/pkg/kv/kvclient/kvstreamer/streamer.go#L346)):*
> ```Go
> 	// The default limit for asynchronous senders.
> 	defaultStreamerConcurrency = 128
> 	// The default scaling factor for the number of streams per vCPU
> ```

Missing punctuation.
___
*[`pkg/kv/kvclient/kvstreamer/streamer.go` line 358 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131226#-O7VUNpL3XHMO1pLWmT9:-O7VUNpL3XHMO1pLWmTA:b-y8dcb8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d8998de678cb3d7bde2cae6fb5303e44c5f05ff/pkg/kv/kvclient/kvstreamer/streamer.go#L358)):*
> ```Go
> 	"kv.streamer.concurrency_limit",
> 	"maximum number of asynchronous requests by a single streamer",
> 	max(defaultStreamerConcurrency, int64(defaultStreamerStreamsPerVCPU*runtime.GOMAXPROCS(0))),
> ```

This is an increase of 32x, not 4x. Do we want to set `defaultStreamerStreamsPerVCPU` to `32` to retain the relationship between this setting and `kv.dist_sender.concurrency_limit`?

**Review by sean- — COMMENTED** [review]
Testing is underway.  I actually think this primitive probably needs to be phased out in favor of other guardrails elsewhere in the system, but raising this here seems like a sensible first start.

I would actually rather see this invert to something like:
```
kv.dist_sender.concurrency_limit = 256*max(2, vCPU)
kv.streamer.concurrency_limit = 128*max(2, vCPU)
```
instead, that way it's always the same multiplier and we just have a floor on the number of vCPUs for the auto-tuning.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131226#-:-O7VXYUw9htsQs550JKC:b-tj9vdy)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2)

**Review by michae2 — COMMENTED** [review]
It would be nice to see how TPC-C and TPC-H change for clusters of different node sizes.

Assuming it's still running, we can keep an eye on the tpch_concurrency roachtest.

> Also, the change only has an effect for nodes with fewer than 16 vCPUs, because the maximum values have not changed. Is that intentionally?

Aren't `defaultSenderConcurrency` and `defaultStreamerConcurrency` minimum values? (Because of the use of `max`.) So I think this PR affects nodes with > 4 vCPUs.

> I would actually rather see this invert to something like:
> 
> ```
> kv.dist_sender.concurrency_limit = 256*min(4, vCPU)
> kv.streamer.concurrency_limit = 128*min(4, vCPU)
> ```

That's fine with me, but I think it would need to be:

```
256*max(4, vCPU)
32*max(4, vCPU)
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131226#-:-O7ViDHZBassZN2H7LW0:bcqwuci)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sean-)

**Review by sean- — COMMENTED** [review]
Updated with new bits that change this to:
```
kv.dist_sender.concurrency_limit = 256*min(2, vCPU)
kv.streamer.concurrency_limit = 32*min(2, vCPU)
```
It may be worth changing the multiplier between dist_sender and streamer to be 1/4 dist_sender, not 1/8, but tbd.  Now things scale purely based on vCPU without arbitrary magic constants.

```
root@localhost:26257/tpcc> show cluster setting kv.streamer.concurrency_limit;
  kv.streamer.concurrency_limit
---------------------------------
                            128
(1 row)

Time: 0ms total (execution 0ms / network 0ms)

root@localhost:26257/tpcc> show cluster setting kv.dist_sender.concurrency_limit;
  kv.dist_sender.concurrency_limit
------------------------------------
                              1024
(1 row)

Time: 0ms total (execution 0ms / network 0ms)

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0          74075          123.5    196.0    142.6    520.1    805.3   1946.2  delivery

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0         741131         1235.2    141.9     96.5    419.4    738.2   3221.2  newOrder

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0          74131          123.6     21.8     16.3     56.6    104.9   1140.9  orderStatus

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0         741265         1235.4    163.4     92.3    570.4    906.0   2952.8  payment

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0          74109          123.5     48.7     24.1    167.8    352.3   1208.0  stockLevel

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
  600.0s        0        1704711         2841.2    144.3     92.3    469.8    805.3   3221.2
Audit check 9.2.1.7: PASS
Audit check 9.2.2.5.1: PASS
Audit check 9.2.2.5.2: PASS
Audit check 9.2.2.5.5: PASS
Audit check 9.2.2.5.6: PASS
Audit check 9.2.2.5.3: PASS
Audit check 9.2.2.5.4: PASS

_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  600.0s    74113.0 5763.1%    141.9     96.5    302.0    419.4    738.2   3221.2
```

```
root@localhost:26257/tpcc> SET CLUSTER SETTING kv.streamer.concurrency_limit=256;
SET CLUSTER SETTING

Time: 8ms total (execution 8ms / network 0ms)

root@localhost:26257/tpcc> SET CLUSTER SETTING kv.dist_sender.concurrency_limit=1024;
SET CLUSTER SETTING

Time: 4ms total (execution 4ms / network 0ms)
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0          78077          130.1    191.8    142.6    503.3    771.8   1811.9  delivery

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0         780801         1301.3    127.8     88.1    369.1    671.1   3623.9  newOrder

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0          78106          130.2     20.4     15.2     56.6    100.7   1208.0  orderStatus

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0         780806         1301.3    136.4     71.3    486.5    805.3   3087.0  payment

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0          78069          130.1     46.9     23.1    159.4    335.5   1208.0  stockLevel

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
  600.0s        0        1795859         2993.1    126.2     79.7    419.4    738.2   3623.9
Audit check 9.2.1.7: PASS
Audit check 9.2.2.5.1: PASS
Audit check 9.2.2.5.2: PASS
Audit check 9.2.2.5.5: PASS
Audit check 9.2.2.5.6: PASS
Audit check 9.2.2.5.3: PASS
Audit check 9.2.2.5.4: PASS

_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  600.0s    78079.5 6071.5%    127.8     88.1    268.4    369.1    671.1   3623.9
```

Literally every RPC-related metric appears to have ticked up slightly







Except for runnable Go routines, which predictably went down because there aren't go routines waiting on the concurrency limit:



Running another test while doubling the values again:

```
root@localhost:26257/tpcc> SET CLUSTER SETTING kv.dist_sender.concurrency_limit=2048;
SET CLUSTER SETTING

Time: 4ms total (execution 4ms / network 0ms)

root@localhost:26257/tpcc> SET CLUSTER SETTING kv.streamer.concurrency_limit=512;
SET CLUSTER SETTING

Time: 4ms total (execution 4ms / network 0ms)

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0          77848          129.7    205.5    151.0    536.9    838.9   2147.5  delivery

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0         778637         1297.7    129.4     88.1    385.9    671.1   2550.1  newOrder

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0          77873          129.8     21.3     15.7     56.6    104.9   1073.7  orderStatus

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0         778560         1297.6    135.2     71.3    486.5    805.3   3087.0  payment

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0          77848          129.7     48.7     24.1    167.8    352.3   1543.5  stockLevel

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
  600.0s        0        1790766         2984.6    127.0     79.7    419.4    738.2   3087.0
Audit check 9.2.1.7: PASS
Audit check 9.2.2.5.1: PASS
Audit check 9.2.2.5.2: PASS
Audit check 9.2.2.5.5: PASS
Audit check 9.2.2.5.6: PASS
Audit check 9.2.2.5.3: PASS
Audit check 9.2.2.5.4: PASS

_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  600.0s    77864.6 6054.8%    129.4     88.1    268.4    385.9    671.1   2550.1
```



I think there's diminishing returns at 8x, but the stddev of RPC heartbeats appears to be smaller, and is allowing for a higher CPU utilization while still keeping the number of queries constant.

My suspicion is increasing this limit is hiding some form of congestion or back-pressure.  Previously, with the smaller settings, any hiccup would cause all pending operations to the node that had a hiccup to backup behind this concurrency limit.  By raising this concurrency limit, any TCP retransmits or transient disk stalls on one node don't block operations to other nodes.

While this runtime concurrency limit is well intentioned and may have helped in the past, a cluster-wide concurrency limit like this is flawed and this concurrency threshold should be per unique sql/kv node.  It is very plausible to me that this concurrency limit is what has led to us observing 0 qps situations in the past from a single sick node.



```
root@localhost:26257/tpcc> SET CLUSTER SETTING kv.streamer.concurrency_limit=384;
SET CLUSTER SETTING

Time: 4ms total (execution 4ms / network 0ms)

root@localhost:26257/tpcc> SET CLUSTER SETTING kv.dist_sender.concurrency_limit=1536;
SET CLUSTER SETTING

Time: 4ms total (execution 4ms / network 0ms)

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0          77419          129.0    212.3    159.4    536.9    838.9   1610.6  delivery

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0         774287         1290.5    131.1     88.1    385.9    671.1   3489.7  newOrder

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0          77439          129.1     21.0     15.7     56.6    104.9   1073.7  orderStatus

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0         774191         1290.3    137.3     75.5    486.5    805.3   2952.8  payment

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0          77371          129.0     49.4     25.2    167.8    352.3   1610.6  stockLevel

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
  600.0s        0        1780707         2967.8    129.0     79.7    436.2    738.2   3489.7
Audit check 9.2.1.7: PASS
Audit check 9.2.2.5.1: PASS
Audit check 9.2.2.5.2: PASS
Audit check 9.2.2.5.5: PASS
Audit check 9.2.2.5.6: PASS
Audit check 9.2.2.5.3: PASS
Audit check 9.2.2.5.4: PASS

_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  600.0s    77428.8 6020.9%    131.1     88.1    285.2    385.9    671.1   3489.7
```

Based off of multiple 10min TPC-C benchmarks, a base value of `384` per vCPU appears to be the sweet spot (up from 64), and a ratio between these settings of 1/4 (up from 1/8).  There may be room for future improvement, but until such time as this limit is removed in favor of something other per-node guardrail, these appear to be good defaults for clusters with at least 8x vCPUs and a 12x node clusters (and are likely good for other clusters and configurations, too).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131226#-:-O7eLZJC9trV9tPqyj5z:bgri993)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sean-)

**Comment by sean-** [comment]
From production in the wild, note the lack of runnable Go routines:
image

**Review by michae2 — COMMENTED** [review]
Nice benchmark! Out of curiosity, how many vCPUs did these nodes have? I have a couple of nits but in general I'm in favor.

> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131226#-:-O7fBbGk38ZTMxItRLv-:b11wo8d)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @sean-)

___
*[`-- commits` line 12 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131226#-O7fAhp33xUEdmqjt_IM):*
nit: to make the release note script work correctly, this needs to be last in the commit message. Also, we should include the details of which settings are being increased and by how much.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131226#-O7f6SKb9oXPiudMrQNv:-O7f6SKb9oXPiudMrQNw:ba0szc2) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a4a64c882ed0cffed03cdc16e695968b4c995f/pkg/kv/kvclient/kvcoord/dist_sender.go#L349)):*
> ```Go
> 	DefaultSenderStreamsPerVCPU = 384
> 	// The minimum number of recommended vCPUs.
> 	MinViableProcs = 2
> ```

By setting this constant to 2, we'll decrease the default for a 2-vCPU node from 1024 to 768. I'm not sure if anyone is actually using 2-vCPU nodes, but WDYT of using 4 for this constant? (We could rename it to something less general if that would help.)
___
*[`pkg/kv/kvclient/kvstreamer/streamer.go` line 344 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131226#-O7f8jr7AZtGs4GkC7Kz:-O7f8jr7AZtGs4GkC7L-:bv8m8mq) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a4a64c882ed0cffed03cdc16e695968b4c995f/pkg/kv/kvclient/kvstreamer/streamer.go#L344)):*
> ```Go
> 
> const (
> 	// The default scaling factor for the number of streams per vCPU.  The value
> ```

nit: I would rewrite this sentence as "The default scaling factor for the number of asynchronous requests per Streamer per vCPU."

**Review by sean- — COMMENTED** [review]
8x vCPUs.  This cluster is a 12x `m6i.xlarge`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131226#-:-O7fKYem7D-nsAdYs8CL:b-ixdocf)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @nvanbenschoten)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131226#-O7f6SKb9oXPiudMrQNv:-O7fLKR4D1tz52tHVmW_:b8xb884) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a4a64c882ed0cffed03cdc16e695968b4c995f/pkg/kv/kvclient/kvcoord/dist_sender.go#L349)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

By setting this constant to 2, we'll decrease the default for a 2-vCPU node from 1024 to 768. I'm not sure if anyone is actually using 2-vCPU nodes, but WDYT of using 4 for this constant? (We could rename it to something less general if that would help.)
</blockquote></details>

Correct.  I thought about this before changing it from `4` to `2`.  The previous values were equally arbitrary, but act as a weird cliff that could hamper CI or other toy installations.  While we do say single proc instances of crdb aren't supported, dual proc is "supported" (but not recommended for prod).

By having this scale per-CPU w/ a floor of 2x vCPUs, we establish our own floor and let customers w/ small deployments have a viable experience.  I'm not opposed to bumping it to 4x, but this seems like a small paper cut  that we could remove from the system while I'm touching these defaults.

**Review by michae2 — APPROVED** [review]
:lgtm:, but let's also wait for @nvanbenschoten.

(Also, don't forget to update the GH pull request description with your new release note.)

> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131226#-:-O7jKMFy1-PhM5YSO8w3:b-9nc51t)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131226#-O7f6SKb9oXPiudMrQNv:-O7jKB6V0RZNcGeZamb-:bb73djm) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a4a64c882ed0cffed03cdc16e695968b4c995f/pkg/kv/kvclient/kvcoord/dist_sender.go#L349)):*
<details><summary><i>Previously, sean- (Sean Chittenden) wrote…</i></summary><blockquote>

Correct.  I thought about this before changing it from `4` to `2`.  The previous values were equally arbitrary, but act as a weird cliff that could hamper CI or other toy installations.  While we do say single proc instances of crdb aren't supported, dual proc is "supported" (but not recommended for prod).

By having this scale per-CPU w/ a floor of 2x vCPUs, we establish our own floor and let customers w/ small deployments have a viable experience.  I'm not opposed to bumping it to 4x, but this seems like a small paper cut  that we could remove from the system while I'm touching these defaults.
</blockquote></details>

SGTM

**Review by nvb — COMMENTED** [review]
> Based off of multiple 10min TPC-C benchmarks, a base value of `384` per vCPU appears to be the sweet spot (up from 64), and a ratio between these settings of 1/4 (up from 1/8).

Doesn't the TPC-C experimentation indicate that 256 per vCPU is better across all latency percentiles?

> Reviewed 1 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131226#-:-O7jgYA9-S0cniz5qOxK:b-ij94om)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @sean-)

**Review by sean- — COMMENTED** [review]
Here's the condensed test results.  I think there's something wrong in the way we do our math, given the identical values for the various p50, p90, and p95 all arriving at the same values.  Instead, I discarded this and focused on the pMax.  I could very easily be convinced that 512/2048 is a better set of defaults given the pMax, but given the pMax, I went with the higher value.

```
128/1024:
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  600.0s    74113.0 5763.1%    141.9     96.5    302.0    419.4    738.2   3221.2

256/1024:
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  600.0s    78079.5 6071.5%    127.8     88.1    268.4    369.1    671.1   3623.9

384/1536:
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  600.0s    77428.8 6020.9%    131.1     88.1    285.2    385.9    671.1   3489.7

512/2048:
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  600.0s    77864.6 6054.8%    129.4     88.1    268.4    385.9    671.1   2550.1
```

Before pushing send on this, I ran a few more tests, and the results (slightly different setup), are interesting:

```
SET CLUSTER SETTING kv.streamer.concurrency_limit=1024;
SET CLUSTER SETTING kv.dist_sender.concurrency_limit=4096;
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  600.	0s    73329.8 5702.2%    149.6    100.7    318.8    436.2    771.8   2818.6

SET CLUSTER SETTING kv.streamer.concurrency_limit=256;
SET CLUSTER SETTING kv.dist_sender.concurrency_limit=1024;
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  600.	0s    72796.1 5660.7%    151.4    100.7    318.8    453.0    771.8   3087.0

SET CLUSTER SETTING kv.streamer.concurrency_limit=1024;
SET CLUSTER SETTING kv.dist_sender.concurrency_limit=4096;
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  600.0s    72369.2 5627.5%    151.8    104.9    318.8    453.0    771.8   2684.4
```

These were 3x 10m tests run back to back.  If anything, I'd ship what's in this PR, or revise and increase this further. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131226#-:-O7jkKn-8hBZHiRwJ3-Z:bngdjji)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @sean-)

**Review by nvb — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131226#-:-O7od_sy5yVmsbNUDhsz:bnfp4nl)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @sean-)

**Comment by sean-** [comment]
image
Based on the above, it may make sense to revise these values from 384/vCPU to 512/vCPU, too.

As an additional baseline, I also benchmarked the following pathologically low values:
```
SET CLUSTER SETTING kv.streamer.concurrency_limit=128;
SET CLUSTER SETTING kv.dist_sender.concurrency_limit=512;
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  600.0s    64318.3 5001.4%    187.0    130.0    402.7    570.4    906.0   3355.4
```
The pMax is interesting in that this value may be influenced by setting this value too low.  It's also interesting that the other percentiles didn't change much, but the average query time did tick up, indicating txns are being queued behind these limits.

At the end of the day, per-node concurrency limits like this are sub-optimal self-protection mechanisms for busy clusters and should be revisited in the future (in favor of per-resource limits, such as per-store and per pair of nodes).  It would not surprise me if we need to double these values again in the next 6-12mo.

**Comment by sean-** [comment]
Final thoughts (the above stats were comparing two tests, not the before and after of this PR):

Summary of Improvements:
	•	Throughput (tpmC) increases by 16.33% with the new settings.
	•	Efficiency (EFC%) improves by 16.33%.
	•	Average Latency decreases by 24.43%.
	•	Median Latency decreases by 25.77%.
	•	Tail Latencies (P90, P95, P99) decrease by 25.02%, 26.48%, and 22.23%, respectively.
	•	Maximum Latency decreases by 16.00%.

TL;DR: we were artificially limiting ourselves (and likely still are elsewhere in the stack) and there's likely further room for improvement for busy clusters.

---
Numerical Differences and Analysis

1.	Throughput (tpmC):
	•	Difference: 74,824.3 - 64,318.3 = +10,506.0
	•	Percentage Increase: (10,506.0 / 64,318.3) × 100% ≈ +16.33%
	•	Analysis: The new settings result in a 16.33% increase in throughput.
2.	Efficiency (EFC %):
	•	Difference: 5,818.4% - 5,001.4% = +817.0%
	•	Percentage Increase: (817.0 / 5,001.4) × 100% ≈ +16.33%
	•	Analysis: Efficiency improves by 16.33%, matching the throughput increase.
3.	Average Latency (Avg(ms)):
	•	Difference: 141.3 ms - 187.0 ms = -45.7 ms
	•	Percentage Decrease: (45.7 / 187.0) × 100% ≈ -24.43%
	•	Analysis: Average latency decreases by 24.43% with the new settings.
4.	Median Latency (P50(ms)):
	•	Difference: 96.5 ms - 130.0 ms = -33.5 ms
	•	Percentage Decrease: (33.5 / 130.0) × 100% ≈ -25.77%
	•	Analysis: Median latency decreases by 25.77%.
5.	90th Percentile Latency (P90(ms)):
	•	Difference: 302.0 ms - 402.7 ms = -100.7 ms
	•	Percentage Decrease: (100.7 / 402.7) × 100% ≈ -25.02%
	•	Analysis: P90 latency decreases by 25.02%.
6.	95th Percentile Latency (P95(ms)):
	•	Difference: 419.4 ms - 570.4 ms = -151.0 ms
	•	Percentage Decrease: (151.0 / 570.4) × 100% ≈ -26.48%
	•	Analysis: P95 latency decreases by 26.48%.
7.	99th Percentile Latency (P99(ms)):
	•	Difference: 704.6 ms - 906.0 ms = -201.4 ms
	•	Percentage Decrease: (201.4 / 906.0) × 100% ≈ -22.23%
	•	Analysis: P99 latency decreases by 22.23%.
8.	Maximum Latency (PMax(ms)):
	•	Difference: 2,818.6 ms - 3,355.4 ms = -536.8 ms
	•	Percentage Decrease: (536.8 / 3,355.4) × 100% ≈ -16.00%
	•	Analysis: Maximum latency decreases by 16.00%.

**Comment by michae2** [comment]
I'm not sure if the backports make sense in this case. Usually we try not to change cluster setting defaults in minor patch releases.
